# GeekBrains, Data Analysis Algorithms
# Lesson 1 Homework

**Импорт библиотек**

In [1]:
import numpy as np
# import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

**Функция ошибки**

In [2]:
def calc_mse(y_true, y_pred):
    return np.mean((y_true - y_pred)**2)

**Данные**

In [3]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # для умножения на intercept
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]]) # стаж
X.shape

(2, 10)

In [4]:
# средний балл ЕГЭ (целевая переменная)
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

## Задание 1
### Условие
Подберите скорость обучения (alpha) и количество итераций:

In [5]:
n = X.shape[1]
alpha = 1e-3
w = np.array([1, 0.5])

for i in range(1000):
    y_pred = np.dot(w, X)
    err = calc_mse(y, y_pred)
    for j in range(w.shape[0]):
        w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err)

0 [1.1102 0.84  ] 3173.15
100 [ 7.43515582 10.71754582] 899.5990444359252
200 [11.44993589 10.49282195] 735.5540619653672
300 [14.98921624  9.8634316 ] 606.0048772266812
400 [18.17511588  9.27626113] 500.7989288428638
500 [21.04606457  8.74617136] 415.355423398327
600 [23.63334499  8.26841272] 345.96206460054725
700 [25.96499229  7.83785632] 289.6039104863001
800 [28.06626432  7.44984037] 243.83236130353217
900 [29.95992301  7.10016181] 206.65877484533297


### Решение Задания 1:

Количество наблюдений и изначальные веса

In [6]:
n = X.shape[1]

Задаем несколько вариантов значений скорости обучения и количества итераций

In [7]:
alphas = [1e-3, 5e-3, 1e-2, 5e-2, 1e-1, 5e-1, 1]
n_iters = np.arange(250, 2001, 250)

Пройдем через все варианты гиперпараметров и посмотрим, какие лучше (некий GridSearch)

In [8]:
results = []
weights_list = {}
for alpha in alphas:
    for n_iter in n_iters:
        # Инициализируем веса
        w = np.array([1, 0.5])
        # Тренируем модель
        for i in range(n_iter):
            y_pred = np.dot(w, X)
            err = calc_mse(y, y_pred)
            for j in range(w.shape[0]):
                w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))
            # if i % 100 == 0:
        print(alpha, n_iter, i, w, err)
        # Записываем результаты
        weights_list[f'w_{alpha}_{n_iter}'] = w
        results.append({"alpha": alpha, "n_iter": n_iter, "error": err})

0.001 250 249 [13.23154215 10.18353219] 668.7033520716412
0.001 500 499 [21.01880845  8.75120433] 416.12485014271545
0.001 750 749 [27.02193294  7.64268415] 265.9861471397773
0.001 1000 999 [31.6502796   6.78802453] 176.7399358190342
0.001 1250 1249 [35.21868711  6.12909079] 123.68974905302211
0.001 1500 1499 [37.96989246  5.62105967] 92.15538726600646
0.001 1750 1749 [40.0910434   5.22937293] 73.41057157477344
0.001 2000 1999 [41.72642888  4.92738649] 62.26818291089486
0.005 250 249 [35.25124915  6.12307796] 123.9168050280122
0.005 500 499 [43.97712557  4.51177813] 51.69332429512342
0.005 750 749 [46.3478067   4.07401372] 46.362350071495655
0.005 1000 999 [46.99188286  3.95508013] 45.96885911973595
0.005 1250 1249 [47.16686806  3.92276778] 45.939814685712896
0.005 1500 1499 [47.21440874  3.91398903] 45.937670852051795
0.005 1750 1749 [47.22732478  3.91160398] 45.93751261096634
0.005 2000 1999 [47.23083386  3.910956  ] 45.93750093084319
0.01 250 249 [43.99927528  4.50768801] 51.6750696

<ipython-input-2-c672c65af498>:2: RuntimeWarning: overflow encountered in square
  return np.mean((y_true - y_pred)**2)
D:\Programs\Anaconda3\envs\ds_projects\lib\site-packages\numpy\core\_methods.py:160: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(arr, axis, dtype, out, keepdims)
D:\Programs\Anaconda3\envs\ds_projects\lib\site-packages\numpy\core\fromnumeric.py:87: RuntimeWarning: overflow encountered in reduce
  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)


0.1 1500 1499 [nan nan] nan
0.1 1750 1749 [nan nan] nan
0.1 2000 1999 [nan nan] nan
0.5 250 249 [-3.02416611e+287 -1.63771504e+288] inf
0.5 500 499 [nan nan] nan


<ipython-input-8-d6dce9a67df8>:12: RuntimeWarning: overflow encountered in multiply
  w[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))


0.5 750 749 [nan nan] nan
0.5 1000 999 [nan nan] nan
0.5 1250 1249 [nan nan] nan
0.5 1500 1499 [nan nan] nan
0.5 1750 1749 [nan nan] nan
0.5 2000 1999 [nan nan] nan
1 250 249 [nan nan] nan
1 500 499 [nan nan] nan
1 750 749 [nan nan] nan
1 1000 999 [nan nan] nan
1 1250 1249 [nan nan] nan
1 1500 1499 [nan nan] nan
1 1750 1749 [nan nan] nan
1 2000 1999 [nan nan] nan


Посмотрим результаты

In [9]:
results

[{'alpha': 0.001, 'n_iter': 250, 'error': 668.7033520716412},
 {'alpha': 0.001, 'n_iter': 500, 'error': 416.12485014271545},
 {'alpha': 0.001, 'n_iter': 750, 'error': 265.9861471397773},
 {'alpha': 0.001, 'n_iter': 1000, 'error': 176.7399358190342},
 {'alpha': 0.001, 'n_iter': 1250, 'error': 123.68974905302211},
 {'alpha': 0.001, 'n_iter': 1500, 'error': 92.15538726600646},
 {'alpha': 0.001, 'n_iter': 1750, 'error': 73.41057157477344},
 {'alpha': 0.001, 'n_iter': 2000, 'error': 62.26818291089486},
 {'alpha': 0.005, 'n_iter': 250, 'error': 123.9168050280122},
 {'alpha': 0.005, 'n_iter': 500, 'error': 51.69332429512342},
 {'alpha': 0.005, 'n_iter': 750, 'error': 46.362350071495655},
 {'alpha': 0.005, 'n_iter': 1000, 'error': 45.96885911973595},
 {'alpha': 0.005, 'n_iter': 1250, 'error': 45.939814685712896},
 {'alpha': 0.005, 'n_iter': 1500, 'error': 45.937670852051795},
 {'alpha': 0.005, 'n_iter': 1750, 'error': 45.93751261096634},
 {'alpha': 0.005, 'n_iter': 2000, 'error': 45.9375009308

По выводу в процессе тренировки можно заметить, что при $ \alpha = 0.05 $, $ n\_iter \geq 500 $ ошибка стабилизируется на значении $ MSE = 45.9375 $. \
В этом случае модель тренируется максимально быстро, сохраняя точность результата. \
Следовательно, параметры $ \alpha = 0.05 $, $ n\_iter = 500 $ можно считать оптимальными для данной модели. \
Также посмотрим веса получившейся модели.

In [10]:
weights_list['w_0.05_500']

array([47.23214286,  3.91071429])

## Задание 2
### Условие
В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее:

In [11]:
alpha = 1e-3

In [12]:
w = np.array([1, 0.5])

for i in range(1000):
    y_pred = np.dot(w, X)
    err = calc_mse(y, y_pred)
    '''for j in range(W.shape[0]):
        W[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))'''
    w -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y))))
    if i % 100 == 0:
        print(i, w, err)

0 [1.4502 0.9502] 3173.15
100 [11.6740207 11.1740207] 737.5884748924025
200 [11.82000082 11.32000082] 737.0919281764903
300 [11.82208519 11.32208519] 737.0918269437198
400 [11.82211495 11.32211495] 737.091826923081
500 [11.82211538 11.32211538] 737.091826923077
600 [11.82211538 11.32211538] 737.0918269230771
700 [11.82211538 11.32211538] 737.091826923077
800 [11.82211538 11.32211538] 737.091826923077
900 [11.82211538 11.32211538] 737.091826923077


### Решение Задания 2:

В коде был забыт аргумент axis=1, который должен был указать, что должны были складываться не все элементы двумерного массива, а только элементы в каждой строке (либо можно было использовать метод np.dot() для произведения матрицы с вектором). \
Получалось, что функция np.sum() считала всего лишь 1 число вместо вектора, что привело к неправильной функции градиента и неверной тренировке.

In [13]:
w = np.array([1, 0.5])

for i in range(1000):
    y_pred = np.dot(w, X)
    err = calc_mse(y, y_pred)
    '''for j in range(W.shape[0]):
        W[j] -= alpha * (1/n * 2 * np.sum(X[j] * (y_pred - y)))'''
#     w -= (alpha * (1/n * 2 * X @ (y_pred - y)))
    w -= (alpha * (1/n * 2 * np.sum(X * (y_pred - y), axis=1)))
    if i % 100 == 0:
        print(i, w, err)

0 [1.1102 0.84  ] 3173.15
100 [ 7.43515582 10.71754582] 899.5990444359252
200 [11.44993589 10.49282195] 735.5540619653672
300 [14.98921624  9.8634316 ] 606.0048772266812
400 [18.17511588  9.27626113] 500.7989288428638
500 [21.04606457  8.74617136] 415.355423398327
600 [23.63334499  8.26841272] 345.96206460054725
700 [25.96499229  7.83785632] 289.6039104863001
800 [28.06626432  7.44984037] 243.83236130353217
900 [29.95992301  7.10016181] 206.65877484533297


## Задание 3
### Условие
Вместо того, чтобы задавать количество итераций, задайте условие остановки алгоритма - когда ошибка за итерацию начинает изменяться ниже определенного порога

### Решение Задания 3:

In [14]:
w = np.array([1, 0.5])
# Задаем минимальное изменение ошибки для продолжения тренировки
min_error_gain = 0.5
# Задаем изначальную шибку как бесконечность
err = np.inf

for i in range(1000):
    y_pred = np.dot(w, X)
    
    # Если изменение ошибки меньше порога, останавливаем тренировку
    prev_err = err
    err = calc_mse(y, y_pred)
    if (prev_err - err) < min_error_gain:
        break
    
    w -= (alpha * (1/n * 2 * X @ (y_pred - y)))
    if i % 100 == 0:
        print(i, w, err)

0 [1.1102 0.84  ] 3173.15
100 [ 7.43515582 10.71754582] 899.5990444359252
200 [11.44993589 10.49282195] 735.5540619653672
300 [14.98921624  9.8634316 ] 606.0048772266812
400 [18.17511588  9.27626113] 500.7989288428639
500 [21.04606457  8.74617136] 415.355423398327
600 [23.63334499  8.26841272] 345.9620646005472
700 [25.96499229  7.83785632] 289.60391048629987
